briefs/m5-thrice.md

```diff
@@ -0,0 +1,187 @@
+
+Python also [specifies an operator precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence). In this section, we will focus on the order functions are called. The ordering described here will help you to deduce the correct answer, but it might not necessarily be the order executed by the Python interpreter.
+
+To a beginner, the order of computation for expressions like those above can be ambiguous. The rule of thumb is that before a function call is made, the function itself and each argument supplied to it needs to be "prepared" first.
+
+Let's see how this works with the two expressions.
+
+### Expression 1
+
+In `thrice(double(add1))(0)`, we are trying to call the function returned by `thrice(double(add1))` with the argument `0`. The argument `0` is already a Python primitive, but `thrice(double(add1))` needs to be evaluated so that it yields a function.
+
+In `thrice(double(add1))`, we are calling `thrice` with the argument `double(add1)`. This time, the function `thrice` does not need to be evaluated because we already know its definition. The argument `double(add1)`, on the other hand, needs to be evaluated.
+
+
+For the function call `double(add1)`, both the function and the argument need not be evaluated, so the call can be made immediately. Once the result of `double(add1)` has been computed, `thrice` can be called with the result of `double(add1)` as its argument. Finally, we can apply the function yielded by `thrice(double(add1))` to `0`.
+
+The order of evaluation can be summarized as
+
+```python
+double(add1)
+thrice(double(add1))
+thrice(double(add1))(0)
+```
+
+### Expression 2
+
+Like in the first expression, we interpret `thrice(double)(add1)(0)` as the function returned by `thrice(double)(add1)` being called with the argument `0`. `thrice(double)(add1)` is interpreted as a call to the function returned by `thrice(double)` with the argument `add1`.
+
+The order of evaluation can be summarized as
+
+```python
+thrice(double)
+thrice(double)(add1)
+thrice(double)(add1)(0)
+```
+
+*"Huh? Why can't we do `(add1)(0)` first for `thrice(double)(add1)(0)`? Doesn't it just return `1`?"*
+
+This seems plausible because in mathematics, the redundant parentheses in `((2*3)+((10/5)))` do not affect the result. They can be removed or ignored. Indeed, this is sometimes true in Python also:
+
+```python
+>>> ((2*3)+((10/5)))
+8.0
+>>> (((max)))(1,9,3)
+9
+>>> (add1)(0)
+1
+```
+
+*"In that case, how do I tell when the parentheses indicate a function call and when I can ignore them? :confused:"*
+
+One heuristic is to look at what precedes the open parenthesis `(`:
```

Comment:
Might be clearer to put quote marks around the ( to show we're referring to the character? But then it won't look like a smiley, haha. your call.